### ISSUE TITLE
06. PR3: projection helper / criterion wrapper の unit test を追加する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#437]
# 06. PR3: projection helper / criterion wrapper の unit test を追加する

## 目的

raw/view projection、Gaussian target、`TraceMemoryConsistencyCriterion` の update order と target selection を unit test で固定する。

この test も SEG-Y / dataset construction に依存しない。small tensor batch を手で作る。

## 依存 issue

- `03_issue_add_projection_and_gaussian_target_helpers.md`
- `04_issue_implement_trace_memory_consistency_criterion.md`
- `05_issue_add_memory_bank_policy_unit_tests.md`

## 対象ファイル

```text
packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_consistency.py
```

同 file に追記してよい。test が長くなりすぎる場合は以下に分けてもよい。

```text
packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_criterion.py
```

## test helper: fake config

PR2 の `CoarseTraceMemoryConsistencyCfg` を直接作る helper を用意する。

```python
def _memory_cfg(**overrides) -> CoarseTraceMemoryConsistencyCfg:
    values = dict(
        enabled=True,
        target_policy=TRACE_MEMORY_TARGET_POLICY_PRIMARY_ANY,
        weight=1.0,
        warmup_batches=0,
        ramp_batches=0,
        ema_decay=0.0,
        update_min_pmax=0.0,
        target_min_pmax=0.0,
        min_count=1,
        max_age_batches=None,
        label_guard_samples=None,
        sigma_ms=None,
    )
    values.update(overrides)
    return CoarseTraceMemoryConsistencyCfg(**values)
```

## test helper: fake batch

最小 batch を手で作る。

```python
def _batch(*, primary=0, secondary=0, raw_ids=None):
    B, H, W = 1, 3, 8
    if raw_ids is None:
        raw_ids = torch.tensor([[10, 11, 12]])
    return {
        'file_index': torch.tensor([0], dtype=torch.long),
        'primary_key_id': torch.tensor([primary], dtype=torch.long),
        'secondary_key_id': torch.tensor([secondary], dtype=torch.long),
        'anchor_raw_indices_view': raw_ids.long(),
        'meta': {
            'trace_valid': torch.ones((B, H), dtype=torch.bool),
            'fb_idx_view': torch.full((B, H), -1, dtype=torch.long),
            'factor': torch.tensor([1.0], dtype=torch.float32),
            'start': torch.tensor([0], dtype=torch.long),
            'dt_eff_sec': torch.tensor([0.001], dtype=torch.float32),
        },
    }
```

必要な test では `trace_tshift_view`, `factor`, `start`, `fb_idx_view` を override する。

## test 1: raw/view projection round-trip

`factor=2.0`, `start=5`, `trace_tshift_view=[[0,1,2]]` のような非自明 case で確認する。

```text
raw -> view -> raw が allclose
```

`trace_tshift_view` が absent の場合も確認する。

## test 2: Gaussian target normalization

```text
centers=[2.0, 5.5]
sigma_bins=[1.0, 2.0]
W=8
```

確認:

```python
assert q.shape == (2, 8)
assert torch.allclose(q.sum(dim=-1), torch.ones(2))
assert q[0].argmax().item() in {2}
assert q[1].argmax().item() in {5, 6}
```

## test 3: first call uses no same-batch memory, then updates memory

base criterion:

```python
def base_criterion(pred, target, batch):
    return get_pick_logits(pred).sum() * 0.0 + torch.tensor(2.0, device=get_pick_logits(pred).device)
```

手順:

```text
criterion = TraceMemoryConsistencyCriterion(... weight=1)
memory initially empty
call once -> total == 2.0, used_rows == 0, memory stored_slot_count > 0
```

これで update order を固定する。loss 計算前に update してしまう実装なら、この test が落ちるようにする。

## test 4: second call can use previous memory target

手順:

```text
1回目: logits の peak を index 3 にする -> memory update
2回目: logits の peak を index 5 にする -> previous memory target index 3 との consistency loss が乗る
```

確認:

```text
total_loss > base_loss
used_rows > 0
```

厳密な loss 値までは固定しなくてよい。finite / positive / used count を確認する。

## test 5: primary_cross は same-primary memory を使わない

手順:

```text
cfg.target_policy = primary_cross
1回目 batch primary=0 -> memory に primary=0 slot が保存される
2回目 batch primary=0 same raw ids -> same-primary なので used_rows == 0
3回目 batch primary=1 same raw ids -> cross-primary なので used_rows > 0
```

この test は今回の実験設計で最重要。same-key target と cross-key target を確実に分ける。

## test 6: secondary_cross は same-primary cross-secondary のみ使う

手順:

```text
cfg.target_policy = secondary_cross
1回目 batch primary=0 secondary=0 -> memory 保存
2回目 batch primary=0 secondary=0 -> same-secondary なので used_rows == 0
3回目 batch primary=0 secondary=1 -> cross-secondary なので used_rows > 0
4回目 batch primary=1 secondary=1 -> cross-primary なので used_rows == 0
```

## test 7: confidence / min_count / warmup / label_guard

まとめて 1 test にしすぎず、失敗原因が分かるように分ける。

最低限:

```text
update_min_pmax が高いと memory が保存されない
target_min_pmax が高いと lookup hit しても used_rows == 0
min_count=2 では 1 回 update だけだと used_rows == 0
warmup_batches > current batch step では total == base_loss だが memory は update される
label_guard_samples が小さいと fb_idx_view と大きくズレた target を拒否する
```

## test 8: no-target case backward

used row が 0 の case で、返った loss に対して `backward()` が通ることを確認する。

```python
logits = torch.randn((1, 1, 3, 8), requires_grad=True)
loss = criterion(logits, target, batch)
loss.backward()
assert logits.grad is not None
```

## 受け入れ条件

- projection / Gaussian / criterion test が単独で通る。
- test は `segyio` に依存しない。
- `primary_cross` と `secondary_cross` の selection semantics が regression として固定される。
- update order が regression として固定される。
- no-target case でも finite loss と backward が成立する。

## 確認コマンド

```bash
pytest packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_consistency.py -q
```

file を分けた場合:

```bash
pytest \
  packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_consistency.py \
  packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_criterion.py \
  -q
```
